src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java

```diff
@@ -96,8 +97,18 @@ public static SearchRequest parseSearchRequest(RestRequest request) {
             }
         }
 
+        // do not allow 'query_and_fetch' or 'dfs_query_and_fetch' search types
+        // from the REST layer. these modes are an internal optimization and should
+        // not be specified explicitly by the user.
+        String searchType = request.param("search_type");
+        if (SearchType.fromString(searchType).equals(SearchType.QUERY_AND_FETCH) ||
+                SearchType.fromString(searchType).equals(SearchType.DFS_QUERY_AND_FETCH)) {
+            throw new ElasticsearchIllegalArgumentException("Unsupported search type [" + searchType + "]");
```

Comment:
> SearchType.fromString is a pure utility function that performs a single, well-defined task:

Maybe we can change its name/contract?

> we might consider adding some function to SearchType to check if it is allowed to be called from public API's, e.g. SearchType.isInternalUseOnly

My preference would go to having two enums: a public one for consumers of our APIs and an internal one to let elasticsearch know how it should execute the query. But I think we should defer it to another issue.